source/Agents/GP/CGPAgent.hpp

```diff
@@ -0,0 +1,118 @@
+/**
+ * This file is part of the Fall 2023, CSE 491 course project.
+ * @brief An Agent based on genetic programming.
+ * @note Status: PROPOSAL
+ **/
+
+#pragma once
+
+#include <cassert>
+#include <iostream>
+#include <map>
+#include <random>
+#include <string>
+
+#include "GPAgent_.hpp"
+#include "GraphBuilder.hpp"
+
+namespace cowboys {
+  /// Don't know the maximum size a state can be, arbitrary large number
+  constexpr size_t INPUT_SIZE = 6;
+
+  /// Number of computational layers for each agent
+  constexpr size_t NUM_LAYERS = 5;
+
+  /// The number of nodes in each layer
+  constexpr size_t NUM_NODES_PER_LAYER = 10;
+
+  /// The number of layers preceding a node's layer that the node can reference
+  constexpr size_t LAYERS_BACK = 2;
+
+  /// @brief An agent based on cartesian genetic programming.
+  class CGPAgent : public GPAgent_ {
+  protected:
+    /// The genotype for this agent.
+    CGPGenotype genotype;
+
+    /// The decision graph for this agent.
+    std::unique_ptr<Graph> decision_graph;
+
+
+
+  public:
+    CGPAgent(size_t id, const std::string &name) : GPAgent_(id, name) {}
+    CGPAgent(size_t id, const std::string &name, const CGPGenotype &genotype)
+        : GPAgent_(id, name), genotype(genotype) {}
+
+
+    void printAgent() override {
```

Comment:
I realize this is an override, but I would capitalize it to match the style of all the other methods (same as serialize below). Assuming there isn't a strong reason to leave them as is.